esp32/machine_timer.c

```diff
@@ -0,0 +1,193 @@
+
+    timer_config_t config;
+    mp_printf(print, "Timer(%p) [", self);
+
+    timer_get_config(self->group, self->index, &config);
+
+    mp_printf(print, "alarm_en=%d ", config.alarm_en);
+    mp_printf(print, "auto_reload=%d ", config.auto_reload);
+    mp_printf(print, "counter_en=%d ", config.counter_en);
+    mp_printf(print, "]");
+}
+
+STATIC mp_obj_t machine_timer_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    mp_arg_check_num(n_args, n_kw, 1, 1, false);
+    machine_timer_obj_t *self = m_new_obj(machine_timer_obj_t);
+    self->base.type = &machine_timer_type;
+
+    self->group = (mp_obj_get_int(args[0]) >> 1) & 1;
+    self->index = mp_obj_get_int(args[0]) & 1;
+
+    return self;
+}
+
+STATIC void machine_timer_disable(machine_timer_obj_t *self) {
+    if (self->handle) {
+        timer_pause(self->group, self->index);
+        esp_intr_free(self->handle);
+        self->handle = NULL;
+    }
+}
+
+STATIC void IRAM_ATTR machine_timer_isr(void *self_in) {
+    machine_timer_obj_t *self = self_in;
+    timg_dev_t *device = self->group?&(TIMERG1):&(TIMERG0);
+
+    device->hw_timer[self->index].update = 1;
+    if (self->index) {
+        device->int_clr_timers.t1 = 1;
+    } else {
+        device->int_clr_timers.t0 = 1;
+    }
+    device->hw_timer[self->index].config.alarm_en = self->repeat;
+
+    mp_sched_schedule(self->callback, self);
+}
+
+STATIC void machine_timer_enable(machine_timer_obj_t *self) {
+    timer_config_t config;
+    config.alarm_en = TIMER_ALARM_EN;
+    config.auto_reload = self->repeat;
+    config.counter_dir = TIMER_COUNT_UP;
+    config.divider = TIMER_DIVIDER;
+    config.intr_type = TIMER_INTR_LEVEL;
+    config.counter_en = TIMER_PAUSE;
+
+    machine_try(timer_init(self->group, self->index, &config));
+    machine_try(timer_set_counter_value(self->group, self->index, 0x00000000ULL));
+    machine_try(timer_set_alarm_value(self->group, self->index, self->period));
+    machine_try(timer_enable_intr(self->group, self->index));
+    machine_try(timer_isr_register(self->group, self->index, machine_timer_isr, (void*)self, TIMER_FLAGS, &self->handle));
+    machine_try(timer_start(self->group, self->index));
+}
+
+STATIC mp_obj_t machine_timer_init_helper(machine_timer_obj_t *self, mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_period,       MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0xffffffff} },
+        { MP_QSTR_mode,         MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 1} },
+        { MP_QSTR_callback,     MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+    };
+
+    machine_timer_disable(self);
+
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // Timer uses an 80MHz base clock, which is divided by the divider/scalar, we then convert to ms.
```

Comment:
That's my code. From the docs, timers always run at APB_CLK_FREQ (80MHz), irrespective of how the processors are clocked. See below on testing.